SensorTag basic board support

ClumsyApe opened the pull request:
Basic board support for the TI SensorTag, with minimal drivers.

* Switch from Gerrit to Github for this changeset
* Split linker script modification and minimal SoC support into 2 commits
* Move uart.h, gpio.h and trng.h to corresponding drivers

Basic board support for the TI SensorTag, with minimal drivers.
Questions remain to be answered, like Stellaris / CC3200 UART driver reuse
and how to accomodate HW-specific flash header / footer needs by modifying
the Cortex generic linker script.

galak reviewed (changes requested):
So the patch series should be re-ordered slightly, the linker script should come first, than SoC (merge the cc2650: Add basic support for Device Tree. patch with SoC patch).

Let's also see about the UART driver.

ClumsyApe commented:
Sorry about that, commit dates mixed up the order when pushing.
Squashed SoC + Device Tree support accordingly.

ClumsyApe commented:
* Had to move gpio.h to arch/ again, because it is also needed by pinmux driver
  (hardware awkardness).
* Tried to use both the stellaris and the CC32xx UART drivers. 
  * With CC32xx UART: TI provides a HAL for CC26xx SoC, "cc26xxware". It is
  different than the HAL used for CC3200 port ("cc32xxware"). Not surprisingly, the CC32xx
  UART driver does not work for the CC2650 (probably power management issues, but see below).
  * With Stellaris UART: Polling interface works, interrupts do not. Pushed code is a quick hack
    to show what's needed to make it work; most importantly, powering on the UART HW module
    used to be done in CC2650 UART driver and was copy-pasted here.
 
I have the following questions:
* What's the policy regarding HALs for Zephyr? I could not find any. If that's preferred,
  I could import TI's cc26xxware HAL along cc32xxware and use it everywhere.
* Regarding SoC power management, should I power on everything that may be needed
  in SoC's init() (in arch/) and not in each individual driver?
* There is still a problem when using Device Tree generated macros: the address is
  baked in their names. Here, a quick hack in the board's .fixup file does the job; but
  why not using generic names? We could generate names using port numbers, for example.

GAnthony commented:
Regarding policy on HALs, my understanding is that they are initially being encouraged as a way to expand board support in Zephyr, though I've heard some comments against HALs if they bring in too much extra code.    In the case of CC32xx, I was able to save ~2K in code space by using the ROM versions of the cc32xx driverlib functions: MAP_*.  
  
I wonder if there is a SimpleLink SDK for cc26xx?  The following link seems to suggest that is the case:
http://dev.ti.com/tirex/content/simplelink_cc13x0_sdk_1_30_00_06/docs/simplelink_mcu_sdk/Users_Guide.html#directory-structure
In which case, cc26xx device files could go under the new ext/ti/hal/simplelink/ directory structure, if you choose to use them.

GAnthony commented:
Regarding power management, the cc32xx UART driver does not currently support power management, so is not a good example.  I would think it's best for the device driver to provide the power management hooks, and initialization.

galak commented on drivers/serial/uart_stellaris.c:
```diff
@@ -247,6 +266,45 @@ static inline void line_control_defaults_set(struct device *dev)
+	reg_write(reg, val);
+}
+
+static void cc2650_prcm_init_uart()
```
could this be done in soc.c for cc2650?

galak commented:
With regards to DT generated, we intend to generate the code bits so the #defines go away, the problem with names is its hard to come up with a generic solution that works for everyone.  Some people start at 0, some want it to be the name the board uses, and not the SoC, etc.

So for now the fixup file exists until we can generate the bits instead that are in the driver code.

ClumsyApe commented:
* Move inline register manipulations functions to soc.c.
* Move power/clock setup of HW modules to soc.c, and remove such code from device drivers, where it was previously located.
* Document the Stellaris UART driver to explain use for CC2650 SoC.

galak commented:
Sorry for taking so long on this, headed on vacation for a few days, but will focus on getting this reviewed and merged next week.  Thanks.

ClumsyApe commented:
No problem, I tried to keep conflicts minimal to avoid headaches upon merging.

galak commented on dts/arm/ti/cc2650.dtsi:
```diff
@@ -0,0 +1,66 @@
+			reg = <0x40001000 0x4C>;
+			interrupts = <5 0>, <6 0>;
+			status = "disabled";
+		};
```
add a label, guessing something like:

label = "UART_0";

galak commented on dts/arm/ti/cc2650.dtsi:
```diff
@@ -53,7 +53,7 @@
 
 		uart0: uart@40001000 {
-			compatible = "ti,cc2650-uart";
+			compatible = "ti,stellaris-uart";
```
make this change in the base SoC patch

galak commented on dts/arm/cc2650_sensortag.fixup:
```diff
@@ -1 +1,7 @@
 #define CONFIG_NUM_IRQ_PRIO_BITS	ARM_V7M_NVIC_E000E100_ARM_NUM_IRQ_PRIORITY_BITS
+
```
add these changes to the board port patch

galak commented on boards/arm/cc2650_sensortag/cc2650_sensortag_defconfig:
```diff
@@ -31,7 +31,8 @@ CONFIG_GPIO_CC2650_INIT_PRIO=0
 # UART driver
```
make these changes in the board port patch

galak commented on drivers/serial/uart_stellaris.c:
```diff
@@ -23,6 +25,7 @@
 #include <init.h>
 #include <uart.h>
 #include <sections.h>
+#include <soc.h>
```
What do we need from soc.h?

galak commented on drivers/serial/uart_stellaris.c:
```diff
@@ -141,6 +144,11 @@ struct uart_stellaris_dev_data_t {
 
+#ifdef CONFIG_SOC_CC2650
+	/* This is normally defined in LM3S6965's soc.h */
+	#define UART_IRQ_FLAGS	0
```
Lets just remove usage of UART_IRQ_FLAGS, just change the function to pass a 0.

galak commented on drivers/pinmux/Kconfig.cc2650:
```diff
@@ -0,0 +1,16 @@
+	bool "Pinmux driver for CC2650 SoC"
+	depends on PINMUX && SOC_SERIES_CC2650
+	select GPIO
+	select GPIO_CC2650
```
since pin mux depends on GPIO driver, the GPIO driver should be first in the sequence of commits.

galak commented on drivers/pinmux/pinmux_cc2650.c:
```diff
@@ -0,0 +1,166 @@
+	REG_ADDR(TI_CC2650_PINMUX_40081000_BASE_ADDRESS, \
+		 CC2650_IOC_IOCFG0 + 0x4 * Func)
+
+static int pinmux_cc2650_init(struct device *dev);
```
you shouldn't need these prototypes, re-order the code in the file if needed so they can be removed.

galak commented on drivers/gpio/gpio_cc2650.c:
```diff
@@ -0,0 +1,348 @@
+#include "gpio_utils.h"
+
+
+#define CC2650_GPIO_IRQ_FLAGS	0
```
remove this, just call function with 0.

galak reviewed (changes requested):
A number of changes needed.  Lets have the 'uart' patch be the first patch, since its just cleaning / commenting on support for CC2650.  Also, lets pull the random # patch from this PR so we can get SoC/Board support in.

galak commented on drivers/random/random_cc2650.c:
```diff
@@ -0,0 +1,215 @@
+ * Texas Instrument's CC2650 SoC.
+ */
+
+#include <toolchain/gcc.h>
```
do we really need toolchain/gcc.h?, trim headres down to what's needed.

ClumsyApe replied:
I'm just using the ARG_UNUSED() macro.

galak replied:
ok

galak commented on drivers/random/random_cc2650.c:
```diff
@@ -0,0 +1,215 @@
+/*
+ * License: Apache-2.0
```
should look like:

 * SPDX-License-Identifier: Apache-2.0

galak reviewed (commented):
some minor comments on random driver.

ClumsyApe commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -41,11 +41,21 @@
+		    CONFIG_TI_CCFG_SIZE)
+
+  #define CCFG_ADDR (ROM_ADDR + ROM_SIZE)
+  #define CCFG_SIZE (CONFIG_TI_CCFG_SIZE)
```
No, I thought initially that other TI devices shared a similar CCFG concept with different options, but I only find this for the CC13xx / CC26xx SoCs. I can remove CONFIG_TI_CCFG_SIZE and hardcode it here, if that's okay.

galak replied:
Please do

ClumsyApe commented on arch/arm/soc/ti_simplelink/cc2650/soc.c:
```diff
@@ -0,0 +1,167 @@
+
+SYS_INIT(ti_cc2650, PRE_KERNEL_1, 0);
+
+u32_t reg_read(u32_t reg)
```
Hadn't seen those, thanks.
I don't get why the implementation seems to be architecture-specific, though (I spot sys_io.h implementations for ARM, Xtensa, and RISC-V). Zephyr only works with 32-bit architectures, so couldn't the implementation be unique?

galak replied:
I'm guessing the implementation are arch specific for any IO specific operations or sync that might be required.

Go ahead and just temporarily use bit_is_set()

ClumsyApe commented:
* Reorganized commits so that GPIO driver comes before pinmux driver,
* Removed random number generator driver, for another pull request,
* Removed CCFG_SIZE config variable, now hard-coded in linker script,
* Various small fixes,
* Fixed license to "SPDX-License-Identifier: Apache-2.0"

galak commented:
Can you rebase the patch set on latest 'arm' branch.  Thanks.

galak commented on arch/arm/soc/ti_simplelink/cc2650/soc.c:
```diff
@@ -0,0 +1,146 @@
+
+SYS_INIT(ti_cc2650, PRE_KERNEL_1, 0);
+
+int bit_is_set(u32_t reg, u8_t bit)
```
nit picking, but we should probably just have bit_is_set as a static inline in the header.

ClumsyApe replied:
Points of view seem to differ on this. I was asked exactly the opposite (implement functions in soc.c separately from soc.h) some weeks ago.

galak commented on dts/arm/ti/cc2650.dtsi:
```diff
@@ -0,0 +1,66 @@
+		};
+
+		uart0: uart@40001000 {
+			compatible = "ti,cc2650-uart";
```
should be: compatible = "ti,stellaris-uart";

galak commented on dts/arm/ti/cc2650.dtsi:
```diff
@@ -0,0 +1,66 @@
+};
+
+&nvic {
+	arm,num-irq-priority-bits = <8>;
```
This doesn't look right, the number of bits should probably be 3?  Guessing 8 priority levels.

ClumsyApe replied:
Good catch, I look like I mixed up the Cortex M3's user guide number of prio levels (256) with the ones available for the CC2650 SoC (8 indeed).

galak reviewed (changes requested):
Some dts related changes need to be made.

galak commented:
I'm going to pull in the linker script and modify the uart driver patch into the arm branch so we can close out on those bits.

ClumsyApe commented:
* Fixed number of priority bits for interrupts (DTSI file)
* Fixed UART driver used (my fault, don't know where I changed the configuration again)

galak commented on drivers/serial/uart_stellaris.c:
```diff
@@ -644,7 +644,7 @@ static struct uart_stellaris_dev_data_t uart_stellaris_dev_data_0 = {
 };
 
-DEVICE_AND_API_INIT(uart_stellaris0, CONFIG_UART_STELLARIS_PORT_0_NAME, &uart_stellaris_init,
+DEVICE_AND_API_INIT(uart_stellaris0, "UART_0", &uart_stellaris_init,
```
Why this change, we should keep things as they are and use label in device tree like other uarts are using.

galak commented on dts/arm/cc2650_sensortag.fixup:
```diff
@@ -1 +1,7 @@
+#define TI_STELLARIS_UART_4000C000_CURRENT_SPEED TI_STELLARIS_UART_40001000_CURRENT_SPEED
+#define UART_STELLARIS_CLK_FREQ  CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC
+#define TI_STELLARIS_UART_4000C000_IRQ_0 TI_STELLARIS_UART_40001000_IRQ_0
+#define TI_STELLARIS_UART_4000C000_IRQ_0_PRIORITY TI_STELLARIS_UART_40001000_IRQ_0_PRIORITY
```
as mentioned before, merge this change into the board patch.

galak reviewed (changes requested):
uart patch still needs work, in theory we should be able to remove it.

ClumsyApe commented:
* Changed Stellaris driver to use DTS-provided labels.
* Moved .fixup changes to board patch.

galak reviewed (changes requested):
the uart: Use DTS labels... commit, should be first, and should also remove the defines from dts/arm/qemu_cortex_m3.fixup.

#define CONFIG_UART_STELLARIS_PORT_0_NAME       TI_STELLARIS_UART_4000C000_LABEL
#define CONFIG_UART_STELLARIS_PORT_1_NAME       TI_STELLARIS_UART_4000D000_LABEL
#define CONFIG_UART_STELLARIS_PORT_2_NAME       TI_STELLARIS_UART_4000E000_LABEL

galak commented on dts/arm/ti/cc2650.dtsi:
```diff
@@ -57,6 +57,7 @@
 			reg = <0x40001000 0x4C>;
 			interrupts = <5 0>, <6 0>;
 			status = "disabled";
+			label = "UART_0";
```
I'd expect this in the SOC commit.

galak commented on samples/drivers/gpio/src/main.c:
```diff
@@ -110,18 +122,27 @@
 #define GPIO_OUT_PIN	2
 #define GPIO_INT_PIN	3
 #define GPIO_NAME	"GPIO_SS_"
+#define GPIO_INT_PIN_EXTRAFLAGS
```
is there really a need to introduce GPIO_INT_PIN_EXTRAFLAGS since its only used in the CC2650 ifdef case?

galak commented on samples/drivers/gpio/src/main.c:
```diff
@@ -163,9 +184,17 @@ void main(void)
+	ret = gpio_pin_configure(gpio_dev, GPIO_INT_PIN,
+				 (GPIO_DIR_IN | GPIO_INT |
+				  GPIO_INT_EDGE | GPIO_INT_ACTIVE_HIGH |
+				  GPIO_INT_DEBOUNCE | GPIO_INT_PIN_EXTRAFLAGS));
```
Why not just replace GPIO_INT_PIN_EXTRAFLAGS with GPIO_PUD_PULL_UP, not sure what the #define is getting us.

ClumsyApe replied:
For the Sensortag, I needed an extra GPIO_PUD_PULL_UP flag.
As this is probably not the only board needing to pass some extra flags, I added a little bit of flexibility, so other boards in the future can define (or not) more flags as needed while keeping code readable. 
So yes, for now it's only used for the CC2650 but it may prove useful as the test code becomes used for more boards.

galak replied:
I tend to do such things once we see the need for them.

ClumsyApe replied:
Alright, I'll just use GPIO_PUD_PULL_UP then.

galak reviewed (changes requested):
Remove GPIO_INT_PIN_EXTRAFLAGS, doesn't seem to add anything.

Commit:
uart: Use DTS labels for Stellaris driver.

Update driver to use DTS-generated #defines for port names,
and not obsolete Kconfig variables.

Signed-off-by: Geoffrey Le Gourriérec <geoffrey.legourrierec@smile.fr>

Commit:
arm: Add support for TI's CC2650 SoC.

Add support in arch/arm/soc/ti_simplelink, along with support
for CC32xx SoC.

Signed-off-by: Geoffrey Le Gourriérec <geoffrey.legourrierec@smile.fr>

Commit:
sensortag: Add TI's SensorTag board.

Add support for TI's SensorTag board, which uses a CC2650 SoC.

Signed-off-by: Geoffrey Le Gourriérec <geoffrey.legourrierec@smile.fr>

Commit:
cc2650: Add GPIO driver.

Signed-off-by: Geoffrey Le Gourriérec <geoffrey.legourrierec@smile.fr>

Commit:
cc2650: Add pinmux driver.

Signed-off-by: Geoffrey Le Gourriérec <geoffrey.legourrierec@smile.fr>

Commit:
samples: gpio: Add support for SensorTag board.

Add extra #defines in samples/drivers/gpio to test
TI SensorTag board.

Signed-off-by: Geoffrey Le Gourriérec <geoffrey.legourrierec@smile.fr>

ClumsyApe commented:
* Moved Stellaris UART driver update commit as first commit.
* Simplify GPIO driver sample code.

ClumsyApe commented:
Thanks a lot Kumar for your work and patience! I'm happy to make my first steps in the Zephyr project. :)

galak commented:
Thanks for your patience as well.  Hope to see some more development from you in the future!
